build.rs

```diff
@@ -66,3 +74,179 @@ fn main() {
     }
     println!("cargo:outdir={}", out_dir);
 }
+
+
+fn build_jsglue() {
+    let out = PathBuf::from(env::var("OUT_DIR").unwrap());
+        
+    cc::Build::new()
+        .flag("-std=c++11")
+        .flag_if_supported("-Wno-unused-parameter")
+        .flag_if_supported("-Wno-invalid-offsetof")
+        .file("src/jsglue.cpp")
+        .include(out.join("dist/include"))
+        .compile("jsglue");
+}
+
+/// Invoke bindgen on the JSAPI headers to produce raw FFI bindings for use from
+/// Rust.
+///
+/// To add or remove which functions, types, and variables get bindings
+/// generated, see the `const` configuration variables below.
+fn build_jsapi_bindings() {
+    let out = PathBuf::from(env::var("OUT_DIR").unwrap());
+
+    // By default, constructors, destructors and methods declared in .h files are inlined,
+    // so their symbols aren't available. Adding the -fkeep-inlined-functions option
+    // causes the jsapi library to bloat from 500M to 6G, so that's not an option.
+    let mut config = bindgen::CodegenConfig::all();
+    config.constructors = false;
+    config.destructors = false;
+    config.methods = false;
+    
+    let mut builder = bindgen::builder()
+        .rust_target(bindgen::RustTarget::Stable_1_19)
+        .header("./src/jsglue.hpp")
+        // Translate every enum with the "rustified enum" strategy. We should
+        // investigate switching to the "constified module" strategy, which has
+        // similar ergonomics but avoids some potential Rust UB footguns.
+        .rustified_enum(".*")
+        .enable_cxx_namespaces()
+        .with_codegen_config(config)
+        .rustfmt_bindings(true)
+        .clang_arg("-I").clang_arg(out.join("dist/include").to_str().expect("UTF-8"))
+        .clang_arg("-x").clang_arg("c++")
+        .clang_arg("-std=c++11")
+        .clang_arg("-fno-sized-deallocation")
+        .clang_arg("-DRUST_BINDGEN");
+
+    if cfg!(feature = "debugmozjs") {
+        builder = builder
+            .clang_arg("-DJS_GC_ZEAL")
+            .clang_arg("-DDEBUG")
+            .clang_arg("-DJS_DEBUG");
```

Comment:
These defines are missing from the jsglue compilation if debugmozjs is enabled. We should probably make some kind of common setup to avoid flags going out of sync.

Reply:
Oh, good point.

Reply:
Rather annoyingly, the flags are set in both `makefile.cargo` and `build.rs`, so we're stuck with some duplication. The thing to do eventually is move everything out of `makefile.cargo`, but I don't think that should be part of this PR.